src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -350,5 +538,142 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
   }
 }
 
+void NSolidHeapSnapshot::sampling_cb(uint64_t thread_id,
+                                     int status,
+                                     const std::string snapshot,
+                                     NSolidHeapSnapshot* snapshotter) {
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+  HeapSamplerStor& stor = it->second;
+  stor.cb(status, snapshot, stor.data.get());
+  if (snapshot.empty()) {
+    snapshotter->threads_running_heap_sampling_.erase(it);
+  }
+}
+
+void NSolidHeapSnapshot::start_sampling_profiler(
+    SharedEnvInst envinst,
+    uint64_t duration,
+    uint64_t snapshot_id,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+
+  HeapSamplerStor& stor = it->second;
+  ASSERT(stor.flags & kFlagIsSamplingHeap);
+
+  v8::Isolate* isolate = envinst->isolate();
+  v8::HeapProfiler* profiler = isolate->GetHeapProfiler();
+  ASSERT_NOT_NULL(profiler);
+
+  profiler->StartSamplingHeapProfiler(
+      stor.sample_interval,
+      stor.stack_depth,
+      stor.sampling_flags);
+
+  // Schedule a timer to take the snapshot
+  int er = QueueCallback(duration,
+                         stop_sampling_profiler,
+                         envinst,
+                         snapshot_id,
+                         snapshotter);
+
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
+  }
+}
+
+void NSolidHeapSnapshot::stop_sampling_profiler(
+    SharedEnvInst envinst,
+    uint64_t snapshot_id,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  if (it == snapshotter->threads_running_heap_sampling_.end()) {
+    // The profiler was stopped before the timer was triggered
+    return;
+  }
+
+  HeapSamplerStor& stor = it->second;
+  if (stor.snapshot_id !=  snapshot_id) {
```

Comment:
Is the linter happy with this?
```suggestion
  if (stor.snapshot_id != snapshot_id) {
```